packages/vite/src/node/environment.ts

```diff
@@ -0,0 +1,31 @@
+import type { DevEnvironment } from './server/environment'
+import type { BuildEnvironment } from './build'
+import type { ScanEnvironment } from './optimizer/scan'
+import type { FutureCompatEnvironment } from './baseEnvironment'
+import type { PluginContext } from './plugin'
+
+export type Environment =
+  | DevEnvironment
+  | BuildEnvironment
+  | ScanEnvironment
+  | FutureCompatEnvironment
```

Comment:
We had a discussions with Pedro about knowing if you are being called during the scan phase [here](https://discord.com/channels/804011606160703521/831456449632534538/1278763329116049450). I think we need to expose an API to know this. The `ScanEnvironment` isn't the same as a `DevEnvironment` because it doesn't expose a `moduleGraph`. I think we can make it part of the API.

About the `FutureCompatEnvironment` and the typing, I think the last one to touch this part was @ArnaudBarre. The idea is to allow plugins to do
```js
if (this.environment.mode === 'dev') {
  this.environment.moduleGraph // this works without a cast
}
```
We could play with a different design though. Maybe if you have an idea here we could check it in a PR to see how it affects `this.environment` usage?

Reply:
IIUC for remix, we don't exactly have to expose `ScanEnvironment`, they can still check with `environment.mode === 'scan'`? (so it's still internal)

I did some test and it seems like we need a kind of `FutureCompatEnvironment` to have type narrowing work 🥲 In that case, what do you think about:

```ts
export type Environment =
  | DevEnvironment
  | BuildEnvironment
  | UnknownEnvironment // `mode: "unknown"`
```

Reply:
`UnknownEnvironment` sounds a lot better to me. I like it. About scan, what would be the type for it, thought?

Reply:
We could still keep `ScanEnvironment` I'd imagine, as long as we don't expose it? Just that if some plugins wants to check if they're in scan mode, they can do:

```ts
// @ts-expect-error
this.environment.mode === 'scan'
```

And they'll not have intellisense of what the ScanEnvironment exposes, and we're free to modify it in the future.

Reply:
Maybe all we need to do is this:
```ts
export type Environment =
  | DevEnvironment
  | BuildEnvironment
  | /** @internal */ ScanEnvironment
  | FutureCompatEnvironment
```
and tweak the dts bundling to strip it.

Reply:
I'm ok with unknown too even if there is a bit more chance people try to do `if (this.environment.mode === 'unknown')` than `if (this.environment.mode === 'future-compat')`

Reply:
I'm not 100% bought into the current `ScanEnvironment`, mainly because if we ever want to have a scan environment during build, we would need to name it `BuildScanEnvironment` (I find the possibility remote given that we tried this direction and backtracked and rolldown is in the picture now, but anyways). So I'm ok making this internal for now.

Reply:
Yeah it is kinda more likely to misunderstood `UnknownEnvironment` and compare with `=== 'unknown'` 🤔 But if we're ok with it, I think it's a little cleaner and we can specifically document to not rely on this.

Reply:
Yeah maybe a bit of documentation on how plugins should check if the right environment to run would be good!

Reply:
Thanks for implementing this Bjorn! https://github.com/vitejs/vite/pull/18008